File: common/src/main/java/dev/wren/truss/platform/services/PlatformHelper.java
package dev.wren.truss.platform.services;

import java.util.Optional;
import java.util.function.Supplier;

public interface PlatformHelper {

    boolean isClient();

    boolean isDedicatedServer();

    String getDist();

    Platform getPlatform();

    enum Platform {
        FORGE,
        FABRIC,
        NEOFORGE
    }

    default void executeWhenOnClient(Supplier<Runnable> toExecute) {
        if (isClient()) toExecute.get().run();
    }

    default void executeWhenOnServer(Supplier<Runnable> toExecute) {
        if (isDedicatedServer()) toExecute.get().run();
    }

    default <T> Optional<T> runWhenOnClient(Supplier<Supplier<T>> toRun) {
        return isClient() ? Optional.of(toRun.get().get()) : Optional.empty();
    }

    default <T> Optional<T> runWhenOnServer(Supplier<Supplier<T>> toRun) {
        return isDedicatedServer() ? Optional.of(toRun.get().get()) : Optional.empty();
    }
}

File: common/src/main/kotlin/dev/wren/truss/platform/services/PlatformHelper.kt
package dev.wren.truss.platform.services

import java.util.Optional
import java.util.function.Supplier


interface PlatformHelper {

    val isClient: Boolean
    val isDedicatedServer: Boolean
    val dist: String
    val platform: Platform

    enum class Platform {
        FORGE,
        FABRIC,
        NEOFORGE
    }

    fun executeWhenOnClient(toExecute: Supplier<Runnable>) {
        if (isClient) toExecute.get().run()
    }

    fun executeWhenOnServer(toExecute: Supplier<Runnable>) {
        if (isDedicatedServer) toExecute.get().run()
    }

    fun <T : Any> runWhenOnClient(toRun: Supplier<Supplier<T>>) =
        if (isClient) Optional.of<T>(toRun.get().get()) else Optional.empty<T>()

    fun <T : Any> runWhenOnServer(toRun: Supplier<Supplier<T>>) =
        if (isDedicatedServer) Optional.of<T>(toRun.get().get()) else Optional.empty<T>()
}